compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -1405,7 +1419,15 @@ trait Implicits:
           if diff < 0 then alt2
           else if diff > 0 then alt1
           else SearchFailure(new AmbiguousImplicits(alt1, alt2, pt, argument), span)
-        case _: SearchFailure => alt2
+        case fail: SearchFailure =>
+          fail.reason match
+            case ambi: AmbiguousImplicits =>
+              if compareAlternatives(ambi.alt1, alt2, disambiguate = true) < 0
```

Comment:
I think this would be the expected behavior for disambiguate 
- when coming from `healAmbiguous` from a standard ambiguity case, to the current search; 
- but not when coming from `healAmbiguous` from an ambiguity which happened in a nested search.

In the latter case, we would then be comparing `alt2` to alternatives for an unrelated expected type.

Reply:
How about we make use disambiguate = true for compareAlternatives when healing nested implicits?

Reply:
As far as I can tell, that would mean potentially faulty as in false positive (but never missing) warnings for a change in resolution in the nested search

Reply:
Perhaps we could do the check twice only in the `3.7-migration` version to keep the better performance in `3.6`

Reply:
Indeed. In light of this, I think we should go with #49. Can you rebase it on the current state?

Reply:
> Can you rebase it on the current state?

Yes, will do that right now + plus first checking if we are in a migration version before the second comparison

Reply:
I've pushed the changes (on #49) after the rebase, the optimization for versions staring from `3.7`, and some documentation.